src/backend/distributed/planner/recursive_planning.c

```diff
@@ -455,6 +466,97 @@ ShouldRecursivelyPlanSubquery(Query *subquery, RecursivePlanningContext *context
 }
 
 
+/*
+ * ShouldRecursivelyPlanSetOperation determines whether the leaf queries of a
+ * set operations tree need to be recursively planned in order to support the
+ * (sub)query as a whole.
+ */
+static bool
+ShouldRecursivelyPlanSetOperation(Query *query, RecursivePlanningContext *context)
+{
+	PlannerRestrictionContext *filteredRestrictionContext = NULL;
+
+	SetOperationStmt *setOperations = (SetOperationStmt *) query->setOperations;
+	if (setOperations == NULL)
+	{
+		return false;
+	}
+
+	if (context->level == 0)
+	{
+		/*
+		 * We cannot push down top-level set operation. Recursively plan the
+		 * leaf nodes such that it becomes a router query.
+		 */
+		return true;
+	}
+
+	if (setOperations->op != SETOP_UNION)
+	{
+		/*
+		 * We can only push down UNION operaionts, plan other set operations
+		 * recursively.
+		 */
+		return true;
+	}
+
+	if (DeferErrorIfUnsupportedUnionQuery(query) != NULL)
+	{
+		/*
+		 * If at least one leaf query in the union is recurring, then all
+		 * leaf nodes need to be recurring.
+		 */
+		return true;
+	}
+
+	filteredRestrictionContext =
+		FilterPlannerRestrictionForQuery(context->plannerRestrictionContext, query);
+	if (!SafeToPushdownUnionSubquery(filteredRestrictionContext))
+	{
+		/*
+		 * The distribution column is not in the same place in all sides
+		 * of the union, meaning we cannot determine distribution column
+		 * equivalence. Recursive planning is necessary.
+		 */
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * RecursivelyPlanSetOperations descends into a tree of set operations
+ * (e.g. UNION, INTERSECTS) and recursively plans all leaf nodes that
+ * contain distributed tables.
+ */
+static void
+RecursivelyPlanSetOperations(Query *query, Node *node,
+							 RecursivePlanningContext *context)
+{
+	if (IsA(node, SetOperationStmt))
+	{
+		SetOperationStmt *setOperations = (SetOperationStmt *) node;
+
+		RecursivelyPlanSetOperations(query, setOperations->larg, context);
+		RecursivelyPlanSetOperations(query, setOperations->rarg, context);
+	}
+	else if (IsA(node, RangeTblRef))
+	{
+		RangeTblRef *rangeTableRef = (RangeTblRef *) node;
+		RangeTblEntry *rangeTableEntry = rt_fetch(rangeTableRef->rtindex,
+												  query->rtable);
+		Query *subquery = rangeTableEntry->subquery;
+
+		if (rangeTableEntry->rtekind == RTE_SUBQUERY &&
+			QueryContainsDistributedTableRTE(subquery))
```

Comment:
Do we actually need `QueryContainsDistributedTableRTE(subquery)` or is it the part you're still thinking about?

Reply:
We don't need to recursively plan subqueries that don't have a distributed table (and things like `SELECT * FROM reference_table LIMIT 10` are already recursively planned at this point).